### PR TITLE
openssl: add extra url to download old versions

### PR DESCRIPTION
**Pull request by davydden**
... and add `1.1.0d`.

p.s. see the available versions at `ftp://openssl.org/source/`.

### CODE REVIEW

**Comment by adamjstewart** [comment]
The old releases are still available at https://www.openssl.org/source/old, maybe we should add a `list_url`? Do we even want to keep old releases available?

**Comment by davydden** [comment]
>  Do we even want to keep old releases available?

i would not bother keeping them. Even if we add `url_for_version`, we would need to constantly update it since something that was available to download in `ftp://openssl.org/source/` is moved to `www.openssl.org/source/old`.

**Comment by adamjstewart** [comment]
> Even if we add url_for_version, we would need to constantly update it since something that was available to download in ftp://openssl.org/source/ is moved to www.openssl.org/source/old.

I don't think that's true. If the version isn't found at ftp://openssl.org/source/, Spack will then try to look in www.openssl.org/source/old

**Comment by eschnett** [comment]
I would keep the old versions around. Otherwise, I might build a software stack one day, and the next day I cannot repeat the build any more -- I have to start all over since OpenSSL has changed under my feet. There's nothing wrong with pushing people towards updating, but forcing people to update from one day to the next is not reasonable. After all, we don't automatically disable existing (old) OpenSSL installs either.

Whatever Spack has built needs to be reproducible.

**Comment by davydden** [comment]
agreed, i will update the PR...

**Comment by davydden** [comment]
@adamjstewart i `grep'`ed `list_url`, but did not see how one would add two locations to look for a download.

**Comment by davydden** [comment]
nevermind, it's actually an extra one to `url`. Will try it now...

**Comment by adamjstewart** [comment]
You don't need two locations, it will default to `url` and if that doesn't work it will spider `list_url`. You'll also probably need to set `list_depth = 2`.

**Comment by adamjstewart** [comment]
There was also a request that we use https in favor of ftp as some sites have security restrictions against ftp. With the following urls:
```python
    homepage = "https://www.openssl.org"                                        
    url      = "https://www.openssl.org/source/openssl-1.1.0d.tar.gz"           
    list_url = "https://www.openssl.org/source/old/"                            
    list_depth = 2         
```
you get:
```
$ spack versions openssl
==> Safe versions (already checksummed):
  1.1.0c  1.0.2i  1.0.2g  1.0.2e  1.0.1u  1.0.1r
  1.0.2j  1.0.2h  1.0.2f  1.0.2d  1.0.1t  1.0.1h
==> Remote versions (not yet checksummed):
  1.1.0d  1.0.1q  1.0.1f  1.0.0q  1.0.0g   0.9.8zf  0.9.8u  0.9.8k  0.9.8a  0.9.7e  0.9.6i
  1.1.0b  1.0.1p  1.0.1e  1.0.0p  1.0.0f   0.9.8ze  0.9.8t  0.9.8j  0.9.8   0.9.7d  0.9.6
  1.1.0a  1.0.1o  1.0.1d  1.0.0o  1.0.0e   0.9.8zd  0.9.8s  0.9.8i  0.9.7m  0.9.7c
  1.1.0   1.0.1n  1.0.1c  1.0.0n  1.0.0d   0.9.8zc  0.9.8r  0.9.8h  0.9.7l  0.9.7b
  1.0.2k  1.0.1m  1.0.1b  1.0.0m  1.0.0c   0.9.8zb  0.9.8q  0.9.8g  0.9.7k  0.9.7a
  1.0.2c  1.0.1l  1.0.1a  1.0.0l  1.0.0b   0.9.8za  0.9.8p  0.9.8f  0.9.7j  0.9.7
  1.0.2b  1.0.1k  1.0.1   1.0.0k  1.0.0a   0.9.8y   0.9.8o  0.9.8e  0.9.7i  0.9.6m
  1.0.2a  1.0.1j  1.0.0t  1.0.0j  1.0.0    0.9.8x   0.9.8n  0.9.8d  0.9.7h  0.9.6l
  1.0.2   1.0.1i  1.0.0s  1.0.0i  0.9.8zh  0.9.8w   0.9.8m  0.9.8c  0.9.7g  0.9.6k
  1.0.1s  1.0.1g  1.0.0r  1.0.0h  0.9.8zg  0.9.8v   0.9.8l  0.9.8b  0.9.7f  0.9.6j
```

**Comment by adamjstewart** [comment]
A demonstration of Spack first checking the default location and then checking the old location:
```
$ spack fetch openssl@0.9.7f
==> Warning: There is no checksum on file to fetch openssl@0.9.7f safely.
  Fetch anyway? [y/N] y
==> Fetching with no checksum.
  Add a checksum or use --no-checksum to skip this check.
==> Fetching https://www.openssl.org/source/openssl-0.9.7f.tar.gz

curl: (22) The requested URL returned error: 404 Not Found
==> Fetching from https://www.openssl.org/source/openssl-0.9.7f.tar.gz failed.
==> Fetching https://www.openssl.org/source/old/0.9.x/openssl-0.9.7f.tar.gz
######################################################################## 100.0%
```

**Comment by davydden** [comment]
@adamjstewart thanks, works indeed!

**Comment by citibeth** [comment]
Let's not run around in circles on this one.  We need to adopt a policy, remember it, and stick to it.  For example:

> There was also a request that we use https in favor of ftp as some sites have security restrictions against ftp.

See #1482, which introduced FTP.

> Even if we add url_for_version

See #1503, which removed `url_for_version()`.

**Comment by adamjstewart** [comment]
To be fair, I removed `url_for_version` because `@system` broke `spack info` and because it wasn't necessary. But `list_url` seems to be all we need here.

I forgot about the switch to ftp for this package. Maybe http > https? I don't know enough about it. @weijianwen would you like to comment? I'm just basing this knowledge off of what you said in #2716:

> compared to ftp, http/https are easier to set up behind a firewall via a transparent proxy

**Comment by davydden** [comment]
@citibeth i put back `ftp` until there is a decision on what to use. However `list_url` is absolutely essential here, without it the current `openssl` is broken altogether since it can not fetch `1.1.0c`.

**Comment by citibeth** [comment]
FYI... I have no opinions on the final answer here.

**Comment by adamjstewart** [comment]
You can actually use ftp for the list_url too.

File: var/spack/repos/builtin/packages/openssl/package.py [hunk 1]
```diff
@@ -33,13 +33,17 @@ class Openssl(Package):
        Layer Security (TLS) and Secure Sockets Layer (SSL) protocols.
        It is also a general-purpose cryptography library."""
     homepage = "http://www.openssl.org"
-    url = "ftp://openssl.org/source/openssl-1.0.1h.tar.gz"
+    url      = "ftp://openssl.org/source/openssl-1.0.1h.tar.gz"
+    list_url = "https://www.openssl.org/source/old/"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/openssl/package.py]: ftp does work, you just need to get rid of the `www.` so that it matches how the url looks.

Comment (davydden) [inline var/spack/repos/builtin/packages/openssl/package.py]: still see some issues
```
==> Fetching from ftp://openssl.org/source/openssl-1.0.2i.tar.gz failed.
==> Could not determine url from list_url.
==> Error: All fetchers failed for openssl-1.0.2i-fhn2p7kuaghzdll3xme4t2zakc3b4uek
```

**Comment by davydden** [comment]
since the last comment is collapsed, here's the current issue with `ftp` in `list_url `:
```
$ spack install openssl@1.0.2i
==> Installing openssl
==> zlib is already installed in /Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/zlib-1.2.10-oh65pcwikiqep74vtfaitp6ojgzivaag
==> perl is already installed in /Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/perl-5.24.0-mhqi2ztwl7zqdh4amffsrpylelk2yjlj
==> Fetching ftp://openssl.org/source/openssl-1.0.2i.tar.gz

curl: (78) RETR response: 550
==> Fetching from ftp://openssl.org/source/openssl-1.0.2i.tar.gz failed.
==> Could not determine url from list_url.
```

File: var/spack/repos/builtin/packages/openssl/package.py (hunk 1 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/openssl/package.py]: Strange, never mind them.

**Comment by davydden** [comment]
@tgamblin i squashed, this should be ready.

**Comment by davydden** [comment]
> Doesn't http require www? Can we use http for both url and list_url?

i think it depends on website and whether or not `www.name.domain` is the same as `name.domain`. But i tried `spack fetch` and the current version (with two `http`) seems to work as expected.

**Comment by adamjstewart** [comment]
Fetch works, but versions doesn't.
```python
list_url = "http://openssl.org/source/old/"
```
```
$ spack versions openssl
==> Safe versions (already checksummed):
  1.1.0c  1.0.2i  1.0.2g  1.0.2e  1.0.1u  1.0.1r
  1.0.2j  1.0.2h  1.0.2f  1.0.2d  1.0.1t  1.0.1h
==> Remote versions (not yet checksummed):
  1.1.0d  1.0.2k
```
```python
list_url = "http://www.openssl.org/source/old/"
```
```
$ spack versions openssl
==> Safe versions (already checksummed):
  1.1.0c  1.0.2i  1.0.2g  1.0.2e  1.0.1u  1.0.1r
  1.0.2j  1.0.2h  1.0.2f  1.0.2d  1.0.1t  1.0.1h
==> Remote versions (not yet checksummed):
  1.1.0d  1.0.2k
```
```python
list_url = "https://openssl.org/source/old/"
```
```
$ spack versions openssl
==> Safe versions (already checksummed):
  1.1.0c  1.0.2i  1.0.2g  1.0.2e  1.0.1u  1.0.1r
  1.0.2j  1.0.2h  1.0.2f  1.0.2d  1.0.1t  1.0.1h
==> Remote versions (not yet checksummed):
  1.1.0d  1.0.2k
```
```python
list_url = "https://www.openssl.org/source/old/"
```
```
$ spack versions openssl
==> Safe versions (already checksummed):
  1.1.0c  1.0.2i  1.0.2g  1.0.2e  1.0.1u  1.0.1r
  1.0.2j  1.0.2h  1.0.2f  1.0.2d  1.0.1t  1.0.1h
==> Remote versions (not yet checksummed):
  1.1.0d  1.0.1q  1.0.1f  1.0.0q  1.0.0g   0.9.8zf  0.9.8u  0.9.8k  0.9.8a  0.9.7e  0.9.6i
  1.1.0b  1.0.1p  1.0.1e  1.0.0p  1.0.0f   0.9.8ze  0.9.8t  0.9.8j  0.9.8   0.9.7d  0.9.6
  1.1.0a  1.0.1o  1.0.1d  1.0.0o  1.0.0e   0.9.8zd  0.9.8s  0.9.8i  0.9.7m  0.9.7c
  1.1.0   1.0.1n  1.0.1c  1.0.0n  1.0.0d   0.9.8zc  0.9.8r  0.9.8h  0.9.7l  0.9.7b
  1.0.2k  1.0.1m  1.0.1b  1.0.0m  1.0.0c   0.9.8zb  0.9.8q  0.9.8g  0.9.7k  0.9.7a
  1.0.2c  1.0.1l  1.0.1a  1.0.0l  1.0.0b   0.9.8za  0.9.8p  0.9.8f  0.9.7j  0.9.7
  1.0.2b  1.0.1k  1.0.1   1.0.0k  1.0.0a   0.9.8y   0.9.8o  0.9.8e  0.9.7i  0.9.6m
  1.0.2a  1.0.1j  1.0.0t  1.0.0j  1.0.0    0.9.8x   0.9.8n  0.9.8d  0.9.7h  0.9.6l
  1.0.2   1.0.1i  1.0.0s  1.0.0i  0.9.8zh  0.9.8w   0.9.8m  0.9.8c  0.9.7g  0.9.6k
  1.0.1s  1.0.1g  1.0.0r  1.0.0h  0.9.8zg  0.9.8v   0.9.8l  0.9.8b  0.9.7f  0.9.6j
```

**Comment by adamjstewart** [comment]
You also need to add `https` to `list_url` if you want `spack versions` to work. I have no idea why it makes a difference, but it does.

**Comment by davydden** [comment]
@adamjstewart thanks, that's an endless PR, it seems :smile:

**Comment by davydden** [comment]
for the record, with current version on the PR, both `1.1.0d` and `1.1.0c`fetch and proceed to stage creation.

**Comment by davydden** [comment]
@tgamblin @becker33 since current version of `openssl` package is plain broken, which means `cmake` (and thereby every package which uses it) is broken for newcomers to Spack, can we please fix this issue by merging the PR?

**Comment by adamjstewart** [comment]
Just needs a quick rebase now that #2958 has been merged.

**Comment by davydden** [comment]
did that, looks reasonable:
```
$ spack install openssl@1.1.0d
==> Installing openssl
==> zlib is already installed in /Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/zlib-1.2.10-oh65pcwikiqep74vtfaitp6ojgzivaag
==> perl is already installed in /Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/perl-5.24.0-mhqi2ztwl7zqdh4amffsrpylelk2yjlj
==> Fetching http://www.openssl.org/source/openssl-1.1.0d.tar.gz
######################################################################## 100.0%
==> Already staged openssl-1.1.0d-2uai7dbubmdcgvn35wm5aj5wks5cdbmq in /Users/davydden/spack/var/spack/stage/openssl-1.1.0d-2uai7dbubmdcgvn35wm5aj5wks5cdbmq
==> No patches needed for openssl
==> Building openssl [Package]
==> Executing phase : 'install'
```

**Pull request merged**